File: packages/runtime-ts/tests/handshake.test.ts
import { describe, it, expect } from 'vitest';
import { SchemaHandshake } from '../src/handshake.js';

describe('SchemaHandshake', () => {
  it('encode/decode roundtrip', () => {
    const hash = new Uint8Array(32).fill(0xab);
    const hs = new SchemaHandshake(hash, '1.2.3');
    const bytes = hs.encode();
    const decoded = SchemaHandshake.decode(bytes);
    expect(Array.from(decoded.hash)).toEqual(Array.from(hash));
    expect(decoded.version).toBe('1.2.3');
  });

  it('check matching hashes', () => {
    const hash = new Uint8Array(32).fill(0x42);
    const local = new SchemaHandshake(hash, '1.0.0');
    const remote = new SchemaHandshake(hash, '1.0.0');
    expect(local.check(remote)).toEqual({ kind: 'match' });
  });

  it('check different hashes', () => {
    const local = new SchemaHandshake(new Uint8Array(32).fill(0x01), '1.0.0');
    const remote = new SchemaHandshake(
      new Uint8Array(32).fill(0x02),
      '1.1.0',
    );
    const result = local.check(remote);
    expect(result.kind).toBe('version_mismatch');
    if (result.kind === 'version_mismatch') {
      expect(result.localVersion).toBe('1.0.0');
      expect(result.remoteVersion).toBe('1.1.0');
    }
  });

  it('wire size is compact', () => {
    const hs = new SchemaHandshake(new Uint8Array(32), '1.0.0');
    const bytes = hs.encode();
    expect(bytes.length).toBe(38);
  });

  it('cross-language wire format', () => {
    const hash = new Uint8Array(32);
    hash[0] = 0x12;
    hash[1] = 0x34;
    const hs = new SchemaHandshake(hash, '2.0.0');
    const bytes = hs.encode();
    expect(bytes[0]).toBe(0x12);
    expect(bytes[1]).toBe(0x34);
    expect(bytes[32]).toBe(5); // LEB128 length of "2.0.0"
    expect(bytes[33]).toBe(0x32); // '2'
  });
});

File: packages/runtime-go/handshake_test.go
package vexil

import (
	"testing"
)

func TestHandshakeEncodeDecodeRoundtrip(t *testing.T) {
	var hash [32]byte
	for i := range hash {
		hash[i] = 0xAB
	}
	hs := NewSchemaHandshake(hash, "1.2.3")
	data := hs.Encode()
	decoded, err := DecodeSchemaHandshake(data)
	assertNoError(t, err)
	if decoded.Hash != hash {
		t.Fatal("hash mismatch")
	}
	if decoded.Version != "1.2.3" {
		t.Fatalf("version mismatch: got %q", decoded.Version)
	}
}

func TestHandshakeCheckMatchingHashes(t *testing.T) {
	var hash [32]byte
	for i := range hash {
		hash[i] = 0x42
	}
	local := NewSchemaHandshake(hash, "1.0.0")
	remote := NewSchemaHandshake(hash, "1.0.0")
	result := local.Check(remote)
	if !result.Match {
		t.Fatal("expected match")
	}
}

func TestHandshakeCheckDifferentHashes(t *testing.T) {
	var h1, h2 [32]byte
	for i := range h1 {
		h1[i] = 0x01
		h2[i] = 0x02
	}
	local := NewSchemaHandshake(h1, "1.0.0")
	remote := NewSchemaHandshake(h2, "1.1.0")
	result := local.Check(remote)
	if result.Match {
		t.Fatal("expected mismatch")
	}
	if result.LocalVersion != "1.0.0" {
		t.Fatalf("local version: got %q", result.LocalVersion)
	}
	if result.RemoteVersion != "1.1.0" {
		t.Fatalf("remote version: got %q", result.RemoteVersion)
	}
}

func TestHandshakeWireSizeCompact(t *testing.T) {
	var hash [32]byte
	hs := NewSchemaHandshake(hash, "1.0.0")
	data := hs.Encode()
	// 32 (hash) + 1 (LEB128 len) + 5 (version) = 38
	if len(data) != 38 {
		t.Fatalf("got %d bytes, want 38", len(data))
	}
}
